src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -2545,7 +2545,7 @@ bool LibraryCallKit::inline_vector_extract() {
         return false;
       }
       // VectorMaskToLongNode requires the input is either a mask or a vector with BOOLEAN type.
-      if (opd->bottom_type()->isa_vectmask() == nullptr) {
+      if (Matcher::mask_op_uses_packed_vector(Op_VectorMaskToLong, opd->bottom_type()->is_vect())) {
```

Comment:
So without this patch, it'd generate `VectorMaskToLong -> URShiftLNode -> AndLNode` (as the earlier `if` condition would have been false) and in the backend, the implementation for `VectorMaskToLong` contains code to convert the mask in a predicate to a packed vector (followed by the actual `VectorMaskToLong` related code). With this patch, it now generates `VectorStoreMaskNode -> VectorMaskToLong -> URShiftLNode ... `(the backend implementation is now separated at the IR level).
Does the major performance uplift come from this Ideal optimization - `VectorMaskToLongNode::Ideal_MaskAll()` where the `VectorStoreMaskNode` gets optimized away?

Reply:
Yes, the IR changes you pointed above is right. 

The major performance uplift comes from the existing optimization  of `VectorStoreMask (VectorLoadMask v) => v`. As you know, `VectorLoadMask` will be generated by some APIs like `VectorMask.fromArray()`. With this change, `VectorMask.fromLong()` also generates this IR. The mask conversions (V->P and P->V) between these APIs can be saved. 

Another performance uplift comes from the flexible vector register allocation. Before, the vector register is specified as the same for different instructions. But now, it depends on RA. In this case, it potentially breaks the un-expected  data-dependence across loop iterations.

Reply:
`VectorStoreMask (VectorLoadMask v) => v` is already existed in C2. Spliting the `VectorLongToMask` and `VectorMaskToLong` can reuse this transformation. That's why the performance can be improved. Because redundent mask conversions are optimized out in some case.